feat(chat): 채팅방 생성 기능 구현

j1suk1m opened the pull request:
## 📌 목적
- 채팅방 생성 기능을 구현합니다.

## ✅ 변경 사항
### 추가된 의존성
- validation
- swagger
- mongodb

### 요청 흐름
```mermaid
sequenceDiagram
    participant Client
    participant Gateway
    participant ChatRoomController
    participant ChatRoomService
    participant MongoDB

    Client->>Gateway: POST /api/chatrooms<br> Body: { name, memberCodes }
    Gateway->>ChatRoomController: 전달 + X-CODE 헤더 삽입
    ChatRoomController->>ChatRoomService: createChatRoom(name, memberCodes, currentMemberCode)
    ChatRoomService->>MongoDB: insert ChatRoom document
    MongoDB-->>ChatRoomService: 저장된 ChatRoom 반환
    ChatRoomService-->>ChatRoomController: ChatRoomCreateResponse 반환
    ChatRoomController-->>Gateway: ResponseDto<ChatRoomCreateResponse> (201 CREATED)
    Gateway-->>Client: 최종 응답 전달

```

## 🧪 테스트 방법
- 서비스 단위 테스트인 ChatRoomServiceTest를 실행합니다.

## 📎 참고 사항
- 계층 구조를 반영하여 엔드포인트를 `/api/chats/`에서 `/api/chatrooms/`로 변경했습니다.

## 📝 제출자 검토 리스트

- [x] 실행 시 오류가 없나요?
- [x] 테스트 전부 통과 했나요?
- [x] 코드 컨벤션을 모두 지켰나요?
- [x] 브랜치가 develop에 merge 요청을 보냈나요?

## 📝 검토자 체크 리스트

> [!IMPORTANT]
> 직접 모든 부분에 대해 검토하고 멘션이나 이모티콘 남겨주세요

- [ ] 요청을 받은 후 하루 이내에 피드백을 보내주세요
- [ ] 본인이 검토해야하는 PR 피드백을 주세요.
- [ ] 브랜치가 develop에 merge 요청을 보냈나요?

j1suk1m committed:
chore(gradle): 의존성 추가

- 유효성 검사를 위한 validation 의존성 추가
- API 문서화를 위한 Swagger 의존성 추가
- 데이터 관리를 위한 MongoDB 의존성 추가

j1suk1m committed:
feat(dto): 채팅방 생성 요청 및 응답 DTO 구현

j1suk1m committed:
feat(response): 공통 응답 구조 구현

- 공통 응답 구조(ResponseDto)에 code, httpStatus, message, data 포함
- 데이터가 없는 경우를 위한 Empty 클래스 구현
- 응답 상태를 정의하는 enum(ResponseDtoStatus) 추가

j1suk1m committed:
feat(exception): 전역 예외 핸들러 구현

- 채팅방 커스텀 예외(ChatRoomException) 추가
- 채팅방 관련 예외 코드 ResponseDtoStatus에 추가
- 전역 예외 핸들러에서 유효성 검사 및 채팅방 커스텀 예외 처리

j1suk1m committed:
feat(chatroom): 채팅방 생성 기능 구현

- 채팅방 엔티티(ChatRoom) 추가
- `@CreatedDate`, `@LastModifiedDate` 사용을 위한 config 추가
- 채팅방 repository, service, controller 추가

j1suk1m committed:
docs(swagger): API 문서화 인터페이스 구현

- ChatRoomControllerApi 생성
- `X-CODE` 헤더는 Gateway에서 처리하므로 Swagger UI에서 숨김

ZZAMBAs commented on communication-service/src/main/java/com/example/communicationservice/repository/ChatRoomRepository.java:
```diff
@@ -0,0 +1,12 @@
+public interface ChatRoomRepository extends MongoRepository<ChatRoom, String> {
+
+    @Query("{ 'memberCodes': { $all: [?0, ?1] } }")
+    boolean existsByMemberCodes(String memberCode1, String memberCode2);
```
`List<String>`으로 받을 수는 없는 걸까요?

j1suk1m replied:
처음에는 `List<String>`으로 받는 방식이 가능한지 몰랐는데,
다시 찾아보니 Spring Data MongoDB에서 지원되는 것을 확인했습니다.
조언해주신 방향으로 반영하겠습니다👍

ZZAMBAs reviewed (approved):
고생하셨습니다. 코드에 따로 큰 문제가 없다고 생각합니다!

sora0319 commented on communication-service/src/main/java/com/example/communicationservice/common/response/Empty.java:
```diff
@@ -0,0 +1,12 @@
+        return INSTANCE;
+    }
+
+}
```
이것을 그대로 사용 시 직렬화 오류가 발생하여서 문자열 반환 getter 같은 변환자가 필요합니다

j1suk1m replied:
태완님 PR을 참고해 해결했습니다. 감사합니다!

sora0319 reviewed (approved):
MongoDB연결 방식에 대해 알 수 있었습니다
수고하셨습니다!

j1suk1m committed:
refactor(entity): LocalDateTime을 Instant로 변경

- 테이블 설계 시 논의했던 사항 반영

j1suk1m committed:
refactor(entity): `@Builder`가 적용된 생성자를 private으로 변경

- 일관된 객체 생성 방식을 사용하기 위해 private 처리

j1suk1m committed:
refactor(chatroom): 채팅방 존재 여부 확인 메서드 시그니처 개선

- 기존: 고정된 참여자 코드만 전달 가능
- 현재: 참여자 코드를 List 형태로 받아 확장성 있고 유연한 조회 가능
- 변경 사항을 서비스 코드에 반영

j1suk1m committed:
fix(service): 채팅방 생성 시 참여자 수 검증 로직 강화

- 기존: 참여자 코드 목록의 크기 확인 -> 중복 코드 포함 가능
- 현재: 스트림을 이용해 참여자 코드 목록의 중복 제거 후 개수 비교

j1suk1m committed:
style(service): 주석 내용 수정

j1suk1m committed:
refactor(service): `@Transactional` 제거

- MongoDB는 단일 document에 대한 쓰기 작업의 원자성을 보장

j1suk1m committed:
fix(response): Empty 클래스 직렬화 오류 해결

j1suk1m committed:
chore(feign): feign client 환경 설정

- feign client 의존성 추가
- 컴포넌트 스캔 패키지 및 활성화 설정

j1suk1m committed:
feat(feign): feign client 인터페이스 및 내부 API 응답 DTO 추가

j1suk1m committed:
feat(exception): 전역 예외 처리 구조 개선

- 공통 예외 구조(CustomException) 추가
- 채팅방 커스텀 예외(ChatRoomException)가 공통 예외 구조를 상속받도록 변경
- 전역 예외 핸들러(GlobalExceptionHandler)가 공통 예외 구조만 처리하도록 개선

j1suk1m committed:
feat(feign): feign 커스텀 예외 처리 구현

j1suk1m committed:
feat(chatroom): 회원 코드 유효성 검증을 위한 feign client 호출 적용

j1suk1m committed:
docs(swagger): 채팅방 생성 API 문서화 수정

- 불필요한 애노테이션 삭제
- 설정 파일 추가

j1suk1m committed:
test(chatroom): 채팅방 생성 로직 서비스 단위 테스트 추가

- 참여자 수, 자기 포함 여부, 유효하지 않은 회원 코드, 중복 채팅방 등 검증
- ResponseDtoStatus 기반 ChatRoomException 발생 여부 검증
- MemberServiceClient, ChatRoomRepository 모킹
